lightstep/resource_metric_dashboard.go

```diff
@@ -650,6 +668,125 @@ func (p *resourceUnifiedDashboardImp) setResourceDataFromUnifiedDashboard(projec
 	return nil
 }
 
+// assembleDashboardPanels takes the incoming set of UnifiedCharts which contain a mix
+// of charts and text panels, then partitions them into separate slices to
+// be processed into distinct Terraform resources.
+func assembleDashboardPanels(
+	dashboardID string,
+	chartSchemaType ChartSchemaType,
+	panels []client.UnifiedChart,
+) (
+	chartResources []interface{},
+	textPanelResources []interface{},
+	err error,
+) {
+	charts := []client.UnifiedChart{}
+	textPanels := []client.UnifiedChart{}
+
+	// Partition by type
+	for _, panel := range panels {
+		switch panel.ChartType {
+		case "timeseries":
+			charts = append(charts, panel)
+		case "text":
+			textPanels = append(textPanels, panel)
+		default:
+			return nil, nil, fmt.Errorf("unknown panel type: %s", panel.ChartType)
+		}
+	}
+
+	chartResources, err = assembleCharts(dashboardID, chartSchemaType, charts)
+	if err != nil {
+		return nil, nil, err
+	}
+	textPanelResources, err = assembleTextPanels(dashboardID, textPanels)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return chartResources, textPanelResources, nil
+}
+
+// assemblePanel copies the data common to all panels (both charts and text panels)
+// into the resource interface
+func setPanelResourceData(
+	resource map[string]interface{}, // Terraform resource
+	panel client.UnifiedChart, // Panel from the API
+) {
+	resource["name"] = panel.Title
+	resource["x_pos"] = panel.Position.XPos
+	resource["y_pos"] = panel.Position.YPos
+	resource["width"] = panel.Position.Width
+	resource["height"] = panel.Position.Height
+
+	resource["id"] = panel.ID
+}
+
+// assembleTextPanels copies the data from the API into the Terraform resource.
+// Note that in Terraform charts and text panels are separate resources, but in
+// the API they are both UnifiedCharts.
+func assembleTextPanels(
+	dashboardID string,
+	textPanels []client.UnifiedChart,
+) ([]interface{}, error) {
+	panelResources := make([]interface{}, 0, len(textPanels))
+	for _, textPanel := range textPanels {
+		if textPanel.ChartType != "text" {
+			return nil, fmt.Errorf("panel %s is not a text panel", textPanel.Title)
+		}
+
+		resource := map[string]interface{}{}
+		setPanelResourceData(resource, textPanel)
+		resource["text"] = textPanel.Text
+		panelResources = append(panelResources, resource)
+	}
+	return panelResources, nil
+}
+
+func assembleCharts(
+	dashboardID string,
+	chartSchemaType ChartSchemaType,
+	chartsIn []client.UnifiedChart,
+) ([]interface{}, error) {
+	var chartResources []interface{}
+	for _, c := range chartsIn {
+		if c.ChartType != "timeseries" {
+			return nil, fmt.Errorf("panel %s is not a timeseries chart", c.Title)
+		}
+
+		resource := map[string]interface{}{}
+		setPanelResourceData(resource, c)
```

Comment:
Moving the common code to `setPanelResourceData` is the only change to `assembleCharts`, otherwise it's just moved code.